jobsdb/internal/cache/cache.go

```diff
@@ -0,0 +1,290 @@
+func filtersToCacheKeys[T ParameterFilter](workspaceFilter string, statesFilter, customValsFilter []string, parametersFilter []T) (workspaceKey string, stateKeys, customValKeys, paramKeys []string) {
+	workspaceKey = workspaceFilter
+	if workspaceFilter == "" { // if no workspace is provided, we use the wildcard
+		workspaceKey = wildcard
+	}
+	stateKeys = statesFilter
+
+	customValKeys = customValsFilter
+	if len(customValKeys) == 0 { // if no custom value is provided, use the wildcard
+		customValKeys = []string{wildcard}
+	}
+	paramKeys = lo.Map(parametersFilter, func(pf T, _ int) string {
+		return fmt.Sprintf("%s:%s", pf.GetName(), pf.GetValue())
+	})
+	if len(paramKeys) == 0 { // if no parameter is provided, we use the wildcard
+		paramKeys = []string{wildcard}
+	}
+	return workspaceKey, stateKeys, customValKeys, paramKeys
+}
+
+// filtersToInvalidationKeys returns the cache keys that need to be invalidated for the provided workspace, states, customVals and parameters filters.
+// Wildcard keys are also returned if needed. An empty slice is returned if all keys need to be invalidated at that level.
+func (c *NoResultsCache[T]) filtersToInvalidationKeys(workspaceFilter string, statesFilter, customValsFilter []string, parametersFilter []T) (workspaceKeys, stateKeys, customValKeys, paramKeys []string) {
+	if workspaceFilter != "" {
+		workspaceKeys = []string{workspaceFilter, wildcard}
+	}
+	stateKeys = statesFilter
+	if len(customValsFilter) > 0 { // include customVals along with the wildcard
+		customValKeys = make([]string, len(customValsFilter)+1)
+		copy(customValKeys, customValsFilter)
+		customValKeys[len(customValsFilter)] = wildcard
+	}
+
+	paramKeys = lo.FilterMap(parametersFilter, func(pf T, _ int) (string, bool) {
+		return fmt.Sprintf("%s:%s", pf.GetName(), pf.GetValue()), lo.Contains(c.supportedParams, pf.GetName())
+	})
+	if len(paramKeys) > 0 { // include params along with the wildcard
+		paramKeys = append(paramKeys, wildcard)
+	}
+	return workspaceKeys, stateKeys, customValKeys, paramKeys
+}
+
+// Print returns a string representation of the cache's tree contents.
+func (c *NoResultsCache[T]) Print() string {
+	if c == nil {
+		return "Cache is nil"
+	}
+	c.cacheTreeMu.RLock()
+	defer c.cacheTreeMu.RUnlock()
+	return fmt.Sprintf("%+v", c.cacheTree)
+}
+
+type cacheEntry struct {
+	noJobs bool
+	tokens []string
+	t      time.Time
+}
+
+// AddToken adds a token to the cache entry and removes the oldest one if there are more than 10.
+func (ce *cacheEntry) AddToken(token string) {
+	ce.tokens = append(ce.tokens, token)
+	if len(ce.tokens) > 10 {
+		ce.tokens = lo.Slice(ce.tokens, 0, 10)
+	}
+}
+
+// SetNoJobs sets the noJobs flag to true if the provided token is found in the cache entry.
+func (ce *cacheEntry) SetNoJobs(token string) bool {
+	for i := len(ce.tokens) - 1; i >= 0; i-- {
+		if ce.tokens[i] == token {
+			ce.noJobs = true
+			ce.t = time.Now()
+			ce.tokens = append(ce.tokens[:i], ce.tokens[i+1:]...)
```

Comment:
[minor] We can use the slices library for this operation if you feel it improves readability

https://pkg.go.dev/golang.org/x/exp/slices#Delete

```suggestion
			ce.tokens = slices.Delete(ce.tokens, i, i+1)
```